link_extractor.py

```diff
@@ -0,0 +1,23 @@
+import requests, urlparse
+from lxml import html
+
+def extract_links(url):
+	try:
+		page = requests.get(url)
+		tree = html.fromstring(page.content)
+		links = tree.xpath('//a/@href')
+		formatted_links = [urlparse.urlparse(link, 'http').geturl() for link in links]
+		return formatted_links
+	except:
+		return []
+
+def valid_links(links, domain):
+	return [link for link in links if is_valid_link(link, domain)]
+
+def is_valid_link(link, domain):
+	if 'www.'+domain in link or link.startswith('mailto:'):
```

Comment:
The logic is to make sure that the link is a valid link to crawl to. Now thinking about it, a false positive is definitely possible if say, domainname.com links to facebook.com/domainname.com or something along those lines. 

The idea behind this method was a link is valid to crawl to if it starts with domainname.com, unless it starts with subdomain.domainname.com, but if a link has www. preceding it we should still allow it (as that isn't a subdomain). The reason why inclusion and not startswith is used is because of http:// and https://, so some valid links don't start with domainname.com or www.domainname.com. Better would be to strip link of http:// or https:// and then use the startswith method.

I tried to use the package urlparse to deal with this, but it ran into problems when no www. was supplied (it would parse domain.com as a path not a network location, so when re-parsing it would give http:///domain.com with three slashes, which web browsers can handle but the Requests package can not).